vllm/v1/engine/core_client.py

```diff
@@ -159,10 +159,16 @@ def __init__(
         atexit.register(self.shutdown)
 
     def shutdown(self):
+        # During final garbage collection in process shutdown, atexit may be
+        # None.
+        if atexit:
+            # in case shutdown gets called via __del__ first
+            atexit.unregister(self.shutdown)
```

Comment:
What is the behavior when we don't do this?

Reply:
- best case `self.shutdown()` gets called more than once, but we handle it gracefully
- if we're doing offline inference and someone creates / deletes LLM instances multiple times, these registered handlers will prevent garbage collection of the objects. Our explicit cleanup calls should still make things work, but it's effectively a memory leak.